### ISSUE TITLE
[RFC] Enhance DX with Turborepo for both Contributors and Adopters

### ISSUE THREAD

**Issue by minkimcello** [backstage/backstage#11570]
**Status:** Open for comments

## Need

The backstage monorepo currently has ~150 packages/plugins and each time you run `yarn build`, it takes anywhere from 40-70 seconds to complete (on an Apple M1 Max). The benefits of a large monorepo unfortunately come at the price of longer build/test times in not only local development but in CI as well. [`Turbo`](https://turborepo.org/) emerged to alleviate some of these pains.

> One thing to keep in mind is that regardless of all the different changes proposed in this RFC, `turbo` (or `nx`) are tools that sit "outside" the monorepo - meaning they can easily be uninstalled and completely removed simply by reverting the scripts at the root `package.json` from `yarn turbo` back to `yarn lerna | workspaces`.

Before we proceed to the proposal, let's first go over some of the immediate benefits `turbo` would provide for backstage:

### `Topological Dependencies`

If you take a look at how `--since` is implemented for [`@backstage/cli`](https://github.com/backstage/backstage/blob/master/packages/cli/src/lib/git.ts#L49-L66) and [`jest`](https://github.com/facebook/jest/blob/main/packages/jest-changed-files/src/git.ts#L53), you'll see they both calculate what's "new" based on `git diff`. That could be an acceptable approach but it should not provide a sense of comprehensiveness as it overlooks all the other packages that are indirectly affected.

For example, let's say we have two packages:
```json
{
  "name": "A",
  "version": "1.0.0",
  "dependencies": {}
}
```
```json
{
  "name": "B",
  "version": "2.0.0",
  "dependencies": {
    "A": "^1.0.0"
  }
}
```
If you introduce new changes in package `B`, you could get away with _just_ testing package `B`. However, a new change introduced to package `A` _should_ result in testing both package `A` _and_ package `B` to ensure there are no breaking changes.

Building and testing every package of your monorepo might be the behaviour you want for CI workflows (say for each merge to the primary branch), but for local development and running tests in CI for pull requests, we could be building and testing packages that are _only_ relevant to the changes of that particular branch.

Turbo can look at the contents of your packages' `package.json` and map out the entire structure of your monorepo. Going back to the example above, if you introduce changes to package `A`, turbo would know that you should also test package `B` since there's something different about package `A`. Furthermore, if you use turbo to build packages `A` and `B`, it would automatically build `A` first before it starts building `B`. (There's a small caveat to this but I'll discuss that in the `Proposal` section.)

### `Task Caching`

In the current state, when you run `yarn build` it will run the build steps from scratch. But theoretically if you run `yarn build` back to back `X` number of times, the results _should_ always be the same as long as there are no new changes to your code. This is where caching comes in real handy. To determine whether or not a task needs to run, Turbo generates hashes based on the contents of your packages, the commands you've executed, and the outputs of those commands.

So say if you had a library that prints "hello" and you build that library, that generates a hash. When you remove the output files of the build and try to run build a second time, turbo will detect that the library itself hasn't changed (resulting in a cache _hit_) so it will "extract" the build files from the previous computation and place them where they should be without having to re-run all the build steps. Then when you change the print value from "hello" to "goodbye" and run build again, the new change will result in a cache _miss_ so it will proceed to actually run the build steps again which then generates a new hash.

Earlier I mentioned how you could instruct turbo to "do TASK for X and all the packages that depend on X", but this requires you to specify `do TASK for X and packages that depend on X` with the [`filter`](https://turborepo.org/docs/core-concepts/filtering) flag. However, because of caching you can instead just say `do X` and turbo will only run the tasks for the relevant packages since all the others will have a cache hit. (It's always a huge plus whenever a developer doesn't need to dig around documentation or the source code to learn all the different CLI flags and options.)

> Where this gets even more interesting is with remote caching - this is when you share your cache with the rest of your team and your CI. You can utilize remote caching for free as an individual, but for a team of developers you need to either [subscribe](https://vercel.com/pricing) to one of their paid plans or implement/deploy your own cache server.

### `Pipeline Configuration`

Another cool feature I wanted to briefly mention is how you can specify dependencies (or "requirements") _of_ your NPM scripts. For instance, if you try to run `yarn build` before ever having run `yarn tsc`, you'll obviously get an error. By configuring `tsc` as a requirement _to_ `build`, you'd never have to run `yarn tsc && yarn build`; you could just run `yarn build`. With the help of caching, turbo will then automatically run `yarn tsc` prior to running `yarn build` if and only when it's necessary.

There are a few other configuration options such as different caching factors or specifying the types of files for calculating if a package should be considered "different" (e.g. if you don't want to re-run tests just because the CHANGELOG.md gets updated), but we won't go into those details here.

## Proposal

- [ ] `Add Turbo`
  - Install `turbo` as a dev dependency for both the create-app template and backstage monorepo and add configurations to `turbo.json`
- [ ] `Update Scripts`
  - Replace `lerna run` with `turbo run` wherever applicable
  - I noticed the NPM scripts could use some tidying up too
    - Some of the scripts are still using deprecated commands and there are some redundant ones (such as the `clean` command configured at the root of the monorepo which essentially just runs `yarn workspaces run clean` twice)
    - When you run `yarn build` for the backstage backend, it also runs the build steps for the app (because the build files for the app get included for the skeleton/bundle of the backend)
      - Without turbo: We could optimize these steps by checking if the app is built (if not throw an error to require the developer to build the app first that way subsequent builds of backends do not need to rebuild the app each time)
        -  The downside to this is it requires developers to remember to rebuild the app whenever there are any changes otherwise the new changes to the frontend won't be included during the image-build
      - With turbo: We could add the build steps of the app as a requirement to building the backend (which, with the caching mechanism, would know when it's necessary to rebuild the app)
- [ ] `Update CI Workflow`
  - The current [CI workflow](https://github.com/backstage/backstage/blob/master/.github/workflows/ci.yml) is configured to run on pull requests and it has the following logic for running build/test:
    ```
    if [ yarn.lock is different than before ]; then
      run build/tests for packages that have direct changes to them
    else
      run build/tests for all packages
    fi
    ```
    I think we can reevaluate and reorganize the CI workflow to be a little more purposeful (regardless of whether or not we use turbo)
      - For a normal pull request, we can run build/test for packages we think are different
      - And for a `Version Packages` Changeset pull request (where all the new changes have been accumulated but not yet published), we can force run all tests
        - I'll create a separate issue for this proposal
- [ ] `Circular Dependencies`
  - In order to use turbo for building our packages and plugins, we _have to_ resolve the circular dependency issue. Unfortunately, that's a hard blocker and there's no way around it
    - Since this won't be a small task, we can keep the build script at root as `yarn workspaces run build` and use turbo once circular dependencies are fixed
- [ ] `Add Volta`
  - Turbo doesn't include your local machine's yarn/node versions when it generates its hash. I'm not sure if there's a reason backstage _can't_ lock in yarn/node versions (I'll create a separate issue for this too), but it's certainly a nice extra safeguard to have. There have been instances where different versions of node would cause errors on one machine and not on the other
- [ ] `Remove Lerna`
  - Lerna just announced two weeks ago that they're transferring their ownership over to `nrwl` but for a while it was not actively maintained. Lerna was useful for a few things that yarn (v1) couldn't do (e.g. running tasks in parallel and maintaining commit histories when you're first converting your polyrepo setup to a monorepo), but with yarn introducing more features and the emergence of new monorepo tools I find lerna has outlived its usefulness in its current state (in my opinion). That being said, I'm not sure what `nrwl` has planned for the future of lerna but unless there's a strong objection I think we can go ahead and get rid of it for now. It'll become one less thing adopters of backstage need to become accustomed to
    - To safely remove lerna:
      - [ ] Double check every instance of where it's used to ensure it can be replaced with `yarn` (or `turbo` if it's at the root); `@backstage/cli` has lerna hard-coded in (I believe for building the app) so that'll also need to be replaced (this is related to the last point of the `Update Scripts` TODO)
- [ ] `Update Docs`
  - [ ] Replace all instances of where there's instructions to use `lerna`
  - [ ] Add section in create-app template README on how to use vercel's remote caching
  - [ ] I can also draft up a guide on how existing backstage adopters can upgrade their backstage workspace to use turbo

## Alternatives

If you've heard of `turbo` before, it's probably safe to assume you've also heard of `nx` - people seem to often compare these two. Before I explain the reason why I concluded turbo to be the better option, here are some of the advantages that `nx` has over `turbo`:

- It's written in typescript whereas `turbo` is primarily written in Go
  - For a large monorepo with specific needs, there may be times when it's necessary to do some source code diving when a tool doesn't behave the way it's expected to
- `nx` has distributed task execution - this is a feature where you can delegate the computations of your tasks to remote systems instead of running them just on your local machine
  - It's not available yet but the creator of `turbo` did say it's something they're currently working on ([source (video)](https://www.youtube.com/watch?v=YX5yoApjI3M))
- And lastly, the most notable feature of `nx` is that they offer public remote caching with the option of generating both read/write and read-only authentication tokens
  - `turbo` originally mentioned that they wanted to offer similar features, but this was before they were acquired by `vercel` and since then there hasn't been any follow up

Despite these selling points for `nx`, the reason why `turbo` was still the obvious choice is that its documentation is very well-written and the API is much more intuitive. The documentation for `nx`, on the other hand, is poorly structured and not properly maintained.

**Comment by Rugvip** [comment]
Thank you for the suggestions! 👍 

It's not the first time `turbo` and `nx` have come up :grin: The first thing I absolutely want to be there is the freedom to pick any monorepo tooling for any Backstage project. I.e. you should be able to switch your own Backstage project to use `turbo` without issues. If there are any issues with that I think this is the best place to focus efforts first.

Looking at the plans for the project and especially what we use in the main repo and ship with `@backstage/create-app`, the main goal right now I'd say is to remove `lerna`. It's really not used for much anymore, with the main thing being `jest` execution because of the memory issues with multi-project configurations. Tbh I think the ideal case here is to be able to run all Jest tests straight from the root with no monorepo orchestration at all.

I'm curious what `turbo` would bring to the table in practice though? For local development you're not exactly running `build` much, it's all about development straight from source. I could see a bit of build caching being useful in CI but tbh I feel the build times are so fast that it's not worth introducing caching at this point, except perhaps for the app build.

Regarding running all tests for `yarn.lock` changes, that's just because we currently aren't able to map lockfile changes to direct dependencies and in the end packages. I feel that's something we could add though x) In the meantime I think it's best to keep running all tests for any lockfile changes though.

**Comment by minkimcello** [comment]
> Looking at the plans for the project and especially what we use in the main repo and ship with @backstage/create-app, the main goal right now I'd say is to remove lerna. It's really not used for much anymore, with the main thing being jest execution because of the memory issues with multi-project configurations. Tbh I think the ideal case here is to be able to run all Jest tests straight from the root with no monorepo orchestration at all.

I'll try to get started on removing lerna this week. 👍 But could you elaborate on what you mean by running all tests straight from the root without monorepo orchestration? I see `backstage-cli test` is being deprecated in favor of `backstage-cli package test` and therefore the new test script should run `yarn workspaces run test` from the root. Were you saying `yarn test` should be doing what `yarn test:all` is currently doing?

> I'm curious what turbo would bring to the table in practice though? For local development you're not exactly running build much, it's all about development straight from source. I could see a bit of build caching being useful in CI but tbh I feel the build times are so fast that it's not worth introducing caching at this point, except perhaps for the app build.

I saw in #8054 that slow build times _is_ an issue for developers which is why I started looking into `turbo` and `nx`. But even if we were to disregard that completely, there are still a few other benefits.

When we set ourselves the goal of using one of these tools, we're forced to tackle a few not-_that_-high-priority-but-still-important chores like moving away from the deprecated backstage-cli commands and removing lerna. But obviously we can do these things without actually adopting a new monorepo tool.

In terms of what kind of benefits we would receive _specifically_ from turbo, I think that boils down to improved reliability and efficiency (even without caching). Whether you're running `build` or `test` (locally or in CI), the question seems to be "should we run this for every package to be extra safe or should we try to run it for only the packages that we think are relevant?". Since turbo can process the relationship between the packages of your monorepo (as opposed to just relying on `git diff`), there's an elevated level of assurance when it tells you "you made changes to this one package to but you should test these other packages too".

Also, the adoption path of turbo will naturally lead to providing more clarity on the backstage-cli commands. I'm not sure if that's on anyone's TODO list, but I think it's important that developers understand what a CLI command does. I've touched up on the build command previously, but if you look at the steps from running `backstage-cli package build` on the backend, it builds the backend package, builds all the other packages, and _then_ bundles the frontend and backend together. Unless the developer dug through the source code of the cli package or notices that some of the terminal output of building the backend also displays the output of when they _only_ build the app, they might assume the build script of the backend _only_ builds the backend.

Instead of relying on `backstage-cli package build` to behave differently based on the role of a package, the command could always just build its own relative package which would allow us to deconstruct the backend build script like this: 
```json
{
  "name": "root",
  "scripts": {
    "build": "build plugins and app and then build backend"
    "build:backend": "yarn build && yarn workspace backend run bundle"
  }
}
```
This would better communicate with developers of the required steps of building the backend.

With turbo we could make that easier to understand (while also utilizing caching) simply by updating the script to:
```json
"build:backend": "turbo run backend#build"
```
And in the configuration say "the `backend#build` requires _all_ the other packages to be built first".

This is one of those one-step-back-two-steps-forward situations and clearly requires more discussion, but I just wanted to point out that the easy-to-read configurations of turbo pipelines (NPM scripts) could allow the backstage-cli package to untangle some of its logic/steps to the monorepo level for the benefit of backstage adopters.

> Regarding running all tests for yarn.lock changes, that's just because we currently aren't able to map lockfile changes to direct dependencies and in the end packages. I feel that's something we could add though x) In the meantime I think it's best to keep running all tests for any lockfile changes though.

Oh for a moment I got confused about the exit code of git diff. I thought it ran all tests if there _weren't_ any lockfile changes. But I think my argument still stands. The lockfile could be the same as before but someone could potentially introduce a new change to a package that could have a side effect to other packages that isn't picked up by the `--since` flag. But yeah, this would be the sort of situation where turbo could be useful. Regardless of the lockfile, it would be able to tell you exactly which packages were changed in the pull request and what other packages are impacted by those changes.

**Comment by juristr** [comment]
Just stumbled upon this :)

I'm part of the Nx & Lerna core team, and we just [released Lerna v6](https://dev.to/nx/lerna-reborn-whats-new-in-v6-245j), which has caching, task pipelines, and more built-in. Seeing that you already have Lerna in use, it might be interesting for you to upgrade (also because it is non-breaking). 

I'm happy to chat/help out if that's desired :)

**Comment by Rugvip** [comment]
@juristr we have actually moved off Lerna. We were only using it for running test and publishing packages previously, but are now running Jest from the root of the repo instead, with multiple projects, which gave us a pretty huge speed increase. Publishing was switched to Yarn 3 `workspaces foreach`. Our development setup is focused heavily around building straight from source, and the production builds are quite close to that too. We basically don't need too much coordination between different package builds.

**Comment by Rugvip** [comment]
Also sorry @minkimcello I must've missed your followup, some good thoughts in there. I totally agree that the CLI as it is structured right now has quite a lot of magic. We moved over to that both because of the optimizations we could do, but also to reduce the number of changes we need to roll out to existing apps.

The commands are generally structured in a way were you can replicate a lower level build setup like the one you describe though, but you'll need to modify the scripts to set that up. You've got the `--skip-build-dependencies` or even the lower level `build-workspace` command. Overall we're aiming for something that works pretty well for most people without modifications, but for those that want to tinker with the build system more it's gonna get more custom and you might use less of the Backstage CLI itself.
